### PR TITLE
New card to encourage anonymous users to become full users

### PR DESCRIPTION
**Pull request by zkhin**
Fixes: https://trello.com/c/KfkYxI8n/13-new-card-for-anonymous-users-to-encourage-them-to-become-full-users

### CODE REVIEW

**Comment by mfogel** [comment]
This looks good. I have no suggestions for improvement. Nice work!

There is one thing we should test manually, at least once. We need to manually end-to-end test the situation where an `ANONYMOUS` user changes to `ACTIVE` and verify the card really does disappear. Have you done this? Here is how I would do it.

This is all with these scripts pointing at your own deployment of the backend. They will use the same `.env` file that the integration tests use.

- use `integration-testing/scripts/sign-user-in-anonymous.js` to create a new a new anonymous user
- use `integration-testing/scripts/run-gql.js` to call the mutation `Mutation.createAnonymousUser`
- use `integration-testing/scripts/run-gql.js` to call the query `Queries.self` and manually verify the new card is there
- use `integration-testing/scripts/run-gql.js` to call the mutation `Mutation.startChangeUserEmail`
  - use an email address like `zayar+test@real.app`
  - receive a 6-digit code from cognito via email
- use `integration-testing/scripts/run-gql.js` to call the mutation `Mutation.finishChangeUserEmail`
- use `integration-testing/scripts/run-gql.js` to call the query `Queries.self` and manually verify the new card has disappeared

Then finally, if you haven't already please make sure all the integration tests pass (not just the new ones). If the test files with GQL subscriptions are failing badly you can skip those files. It seems that some days GQL subscriptions work ok, and other days not so much... I assume it's dependent on network conditions.

Let me know if you get stuck doing these and I can help. Cheers & thanks

**Comment by zkhin** [comment]
> This looks good. I have no suggestions for improvement. Nice work!
> 
> There is one thing we should test manually, at least once. We need to manually end-to-end test the situation where an `ANONYMOUS` user changes to `ACTIVE` and verify the card really does disappear. Have you done this? Here is how I would do it.
> 
> This is all with these scripts pointing at your own deployment of the backend. They will use the same `.env` file that the integration tests use.
> 
> * use `integration-testing/scripts/sign-user-in-anonymous.js` to create a new a new anonymous user
> * use `integration-testing/scripts/run-gql.js` to call the mutation `Mutation.createAnonymousUser`
> * use `integration-testing/scripts/run-gql.js` to call the query `Queries.self` and manually verify the new card is there
> * use `integration-testing/scripts/run-gql.js` to call the mutation `Mutation.startChangeUserEmail`
>   
>   * use an email address like `zayar+test@real.app`
>   * receive a 6-digit code from cognito via email
> * use `integration-testing/scripts/run-gql.js` to call the mutation `Mutation.finishChangeUserEmail`
> * use `integration-testing/scripts/run-gql.js` to call the query `Queries.self` and manually verify the new card has disappeared
> 
> Then finally, if you haven't already please make sure all the integration tests pass (not just the new ones). If the test files with GQL subscriptions are failing badly you can skip those files. It seems that some days GQL subscriptions work ok, and other days not so much... I assume it's dependent on network conditions.
> 
> Let me know if you get stuck doing these and I can help. Cheers & thanks

@mfogel Thanks, I will figure out soon

**Comment by zkhin** [comment]
@mfogel I followed your steps and tested on `run-gql.js` manually and it worked as we expected. 👍

File: real-main/app_tests/models/card/test_manager_on.py [hunk 1]
```diff
@@ -512,3 +512,21 @@ def test_on_user_change_update_photo_card_scenario2(card_manager, user):
...
+    # create ANONYMOUS user, check card is created
+    user.item['userStatus'] = UserStatus.ANONYMOUS
+    card_manager.on_user_change_update_anonymous_upsell_card(user.id, new_item=user.item)
+    assert card_manager.get_card(template.card_id)
+
+    # modify user status to ACTIVE, process, check card is deleted
+    old_item = user.item.copy()
+    user.item['userStatus'] = UserStatus.ACTIVE
+    card_manager.on_user_change_update_anonymous_upsell_card(user.id, new_item=user.item, old_item=old_item)
+    assert card_manager.get_card(template.card_id) is None
```
Comment (mfogel) [inline real-main/app_tests/models/card/test_manager_on.py:532]: Could you please add a stanza to test that creating a user in `ACTIVE` state does not generate a card?

**Review by mfogel — CHANGES_REQUESTED** [review]
Ok, I found something minor. Please address the comment above.

Thanks for doing that manual end-to-end test.

When running the integration tests on this branch, I'm getting a failure in `tests/cards/add-profile-photo.test.js`. I think a test in there is affected by this change and needs to be updated. Could you take a look? Thank you

File: real-main/app_tests/models/card/test_manager_on.py (hunk 1 above)
Comment (zkhin) [reply real-main/app_tests/models/card/test_manager_on.py:532]: Sure

**Pull request merged**